File: web-ts/src/integers/GreaterThan.ts
/**
 * GreaterThan - Integer comparison predicate.
 * @JAVA_REF org.modelingvalue.nelumbo.integers.GreaterThan
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Predicate } from '../logic/Predicate';
import type { Functor } from '../patterns/Functor';
import type { InferContext } from '../InferContext';
import type { InferResult } from '../InferResult';
import { NInteger } from './NInteger';

export class GreaterThan extends Predicate {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, args[0], args[1]);
  }

  protected static fromDataGreaterThan(data: unknown[], declaration?: Predicate): GreaterThan {
    const node = Object.create(GreaterThan.prototype) as GreaterThan;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Predicate | undefined })._declaration = declaration ?? node;
    (node as any)._binding = null;
    (node as any)._hashCodeCached = false;
    (node as any)._hashCode = 0;
    (node as any)._nrOfUnbound = -1;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Predicate): Predicate {
    return GreaterThan.fromDataGreaterThan(data, declaration ?? this.declaration());
  }

  // @JAVA_REF org.modelingvalue.nelumbo.integers.GreaterThan#infer(int, InferContext)
  protected override inferInternal(_nrOfUnbound: number, _context: InferContext): InferResult {
    const nrOfUnbound = this.nrOfUnbound();
    if (nrOfUnbound > 1) {
      return this.unresolvable();
    }

    const l = this.getIntegerVal(0);
    const r = this.getIntegerVal(1);

    if (l === null) {
      return (this.set(0, this.get(1)) as GreaterThan).falsehoodsII();
    }
    if (r === null) {
      return (this.set(1, this.get(0)) as GreaterThan).falsehoodsII();
    }

    return l > r ? this.factCC() : this.falsehoodCC();
  }

  private getIntegerVal(index: number): bigint | null {
    const val = this.get(index);
    if (val instanceof NInteger) {
      return val.value();
    }
    return null;
  }
}

File: src/main/java/org/modelingvalue/nelumbo/rationals/GreaterThan.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.rationals;

import java.io.Serial;
import java.math.BigInteger;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.NelumboConstructor;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.logic.InferContext;
import org.modelingvalue.nelumbo.logic.InferResult;
import org.modelingvalue.nelumbo.logic.Predicate;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class GreaterThan extends Predicate {
    @Serial
    private static final long serialVersionUID = -9139221151771172295L;

    @NelumboConstructor
    public GreaterThan(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args[0], args[1]);
    }

    private GreaterThan(Object[] array, List<AstElement> elements, GreaterThan declaration) {
        super(array, elements, declaration);
    }

    @Override
    protected GreaterThan struct(Object[] array, List<AstElement> elements, Node declaration) {
        return new GreaterThan(array, elements, (GreaterThan) declaration);
    }

    @Override
    protected InferResult infer(int nrOfUnbound, InferContext context) {
        if (nrOfUnbound > 1) {
            return unresolvable();
        }

        BigInteger ln = getVal(0, 0);
        BigInteger ld = getVal(0, 1);
        BigInteger rn = getVal(1, 0);
        BigInteger rd = getVal(1, 1);
        if (ln == null) {
            return set(0, get(1)).falsehoodsII();
        }
        if (rn == null) {
            return set(1, get(0)).falsehoodsII();
        }
        return ln.multiply(rd).compareTo(ld.multiply(rn)) > 0 ? factCC() : falsehoodCC();
    }

}
